README.md
# enka
_Enkapsulate_ tour date!

_enka_ is a tool to encrypt strings to a format, which can be decrypted by other tools programs too.

## The problem
Encrypting data requires knowledge of certain parameters like the encryption algorithm etc.
OpenSSL does not include such information in its output. So a solution is needed.

## The solution
_enka_ steps in and provided a solution.

```
$ ./enka encrypt --algo aes256cbc -kdf pbkdf2:650000:sha256 --text kasnudelisbesseralswienerschnitzel --key obaehrlich
%enka%v1%aes256cbc%pbkdf2:650000:sha256%DH3Z4u0DeLk=%RoTCnO59r8pe2PeJDJoh1Q==%sWyo2KkRVk4PfgWAx/OeYb46SeDXwn/pUhOn2/6rAKXBr/w233cfavspr0GThMXP
$ 
```
This can be dissected into following parts:  
```
                program enka
                version v1
              algorithm aes256cbc
key derivation function pbkdf2:650000:sha256 (function : iteration count : hash function)
                   salt DH3Z4u0DeLk=
   intialization vector RoTCnO59r8pe2PeJDJoh1Q==
         encrypted text sWyo2KkRVk4PfgWAx/OeYb46SeDXwn/pUhOn2/6rAKXBr/w233cfavspr0GThMXP
```

With that information the text can be decrypted by other programs:
```
$ echo $(echo -n "sWyo2KkRVk4PfgWAx/OeYb46SeDXwn/pUhOn2/6rAKXBr/w233cfavspr0GThMXP" | openssl enc -d -base64 -A -aes-256-cbc -pbkdf2 -iter 650000 -md sha256 -S 0c7dd9e2ed0378b9 -iv 4684c29cee7dafca5ed8f7890c9a21d5 -pass pass:obaehrlich)
kasnudelisbesseralswienerschnitzel
$ 
```

But using the _enka_ you can get to your result more quickly:
```
$ echo $(./enka decrypt --string %enka%v1%aes256cbc%pbkdf2:650000:sha256%DH3Z4u0DeLk=%RoTCnO59r8pe2PeJDJoh1Q==%sWyo2KkRVk4PfgWAx/OeYb46SeDXwn/pUhOn2/6rAKXBr/w233cfavspr0GThMXP --key obaehrlich)
kasnudelisbesseralswienerschnitzel
$ 
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- basic command line interface
- encrypt via `encrypt` subcommand
- decrypt via `decrypt` subcommand
- test encryption and decryption

[Unreleased]: https://github.com/joestr/enka/compare/87ea2ab7...HEAD


main_test.go
//
// Copyright (c) 2023 Joel Strasser <joelstrasser1@gmail.com>
//
// Licensed under the EUPL-1.2 license.
//
// For the full license text consult the 'LICENSE' file from the repository.
//

package main

import (
	"enka/enka_decryption"
	"enka/enka_encryption"
	"testing"
)

func TestDefaultEncryption(t *testing.T) {
	args := []string{"--algo", "aes256cbc", "--kdf", "pbkdf2:650000:sha256", "--key", "1234", "--text", "abcd"}
	enka_encryption.Encrypt(args, nil, nil)
}

func TestDefaultDecryption(t *testing.T) {
	args := []string{"--string", "%enka%v1%aes256cbc%pbkdf2:650000:sha256%rhGye/PFC9U=%GiGjGRW2BrQxYdiY69Dstg==%BVeCM81SXQqSwVuMjRyIIQ==", "--key", "1234"}
	enka_decryption.Decrypt(args, nil, nil)
}


main.go
//
// Copyright (c) 2023 Joel Strasser <joelstrasser1@gmail.com>
//
// Licensed under the EUPL-1.2 license.
//
// For the full license text consult the 'LICENSE' file from the repository.
//

package main

import (
	"enka/enka_decryption"
	"enka/enka_encryption"
	"flag"
	"fmt"
	"log"
	"os"
)

var commands = map[string]func([]string, *log.Logger, *log.Logger){
	"encrypt": enka_encryption.Encrypt,
	"decrypt": enka_decryption.Decrypt,
}

var help bool
var errorLog = log.New(os.Stderr, "[enka] ", 0)
var outLog = log.New(os.Stdout, "[enka] ", 0)

func main() {
	flag.BoolVar(&help, "help", false, "Show the global help")
	flag.Parse()

	if len(flag.Args()) == 0 {
		globalUsage(os.Args[0])
		globalHelp()
		os.Exit(1)
	}

	command, ok := commands[flag.Args()[0]]
	if !ok {
		globalUsage(os.Args[0])
		os.Exit(1)
	}

	command(flag.Args()[1:], outLog, errorLog)
}

func globalUsage(executable string) {
	fmt.Printf("Usage: %s [ --help ] encrypt [ --algo ] [ --kdf ] [ --key ] [ --salt ] [ --text ] [ --verbose ] | decrypt [ --key ] [ --string ]\n", executable)
}

func globalHelp() {
	fmt.Println("Help:")
	fmt.Println("  encrypt      The encryption module")
	fmt.Println("  decrypt      The decryption module")
}


pkcs7/pkcs7.go
//
// Copyright (c) 2023 Joel Strasser <joelstrasser1@gmail.com>
//
// Licensed under the EUPL-1.2 license.
//
// For the full license text consult the 'LICENSE' file from the repository.
//

package pkcs7

import (
	"bytes"
	"errors"
	"fmt"
)

// Pkcs7strip remove pkcs7 padding
func Pkcs7strip(data []byte, blockSize int) ([]byte, error) {
	length := len(data)
	if length == 0 {
		return nil, errors.New("pkcs7: Data is empty")
	}
	if length%blockSize != 0 {
		return nil, errors.New("pkcs7: Data is not block-aligned")
	}
	padLen := int(data[length-1])
	ref := bytes.Repeat([]byte{byte(padLen)}, padLen)
	if padLen > blockSize || padLen == 0 || !bytes.HasSuffix(data, ref) {
		return nil, errors.New("pkcs7: Invalid padding")
	}
	return data[:length-padLen], nil
}

// Pkcs7pad add pkcs7 padding
func Pkcs7pad(data []byte, blockSize int) ([]byte, error) {
	if blockSize <= 1 || blockSize >= 256 {
		return nil, fmt.Errorf("pkcs7: Invalid block size %d", blockSize)
	} else {
		padLen := blockSize - len(data)%blockSize
		padding := bytes.Repeat([]byte{byte(padLen)}, padLen)
		return append(data, padding...), nil
	}
}


enka_decryption/decryption.go
//
// Copyright (c) 2023 Joel Strasser <joelstrasser1@gmail.com>
//
// Licensed under the EUPL-1.2 license.
//
// For the full license text consult the 'LICENSE' file from the repository.
//

package enka_decryption

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/base64"
	"encoding/hex"
	"enka/pkcs7"
	"flag"
	"fmt"
	"golang.org/x/crypto/pbkdf2"
	"hash"
	"log"
	"strconv"
	"strings"
)

var enkaString string
var encryptionKey string
var verbose bool

func Decrypt(args []string, outLog *log.Logger, errorLog *log.Logger) {
	fs := flag.NewFlagSet("decrypt", flag.ExitOnError)
	fs.StringVar(&enkaString, "string", "", "The 'enka' string to parse and decrypt")
	fs.StringVar(&encryptionKey, "key", "", "The decryption key")
	fs.BoolVar(&verbose, "verbose", false, "Set verbosity on")
	err := fs.Parse(args)
	if err != nil {
		return
	}

	var splittedEnkaString = strings.Split(enkaString, "%")

	if len(splittedEnkaString) == 0 {
		errorLog.Fatalln("String has an invalid format")
	}

	if splittedEnkaString[1] != "enka" {
		errorLog.Fatalln("String is not an 'enka' string")
	}

	if splittedEnkaString[2] != "v1" {
		errorLog.Fatalln("This version of an 'enka' string is not supported")
	}

	if len(splittedEnkaString[3:]) != 5 {
		errorLog.Fatalln(fmt.Sprintf("Five sections are required; got %s", len(splittedEnkaString[2:])))
	}

	var encryptionKeyBytes = []byte(encryptionKey)
	var encryptionAlgorithm = splittedEnkaString[3]
	var keyDerivationFunction = splittedEnkaString[4]
	var encryptionKeySaltBytes, _ = base64.RawStdEncoding.DecodeString(strings.Replace(splittedEnkaString[5], "=", "", -1))
	var initializationVector, _ = base64.RawStdEncoding.DecodeString(strings.Replace(splittedEnkaString[6], "=", "", -1))
	var cipherBytes, _ = base64.RawStdEncoding.DecodeString(strings.Replace(splittedEnkaString[7], "=", "", -1))

	if !isSupportedAlgo(encryptionAlgorithm) {
		errorLog.Fatalln(fmt.Sprintf("The specified algorithm \"%s\" is not supported", encryptionAlgorithm))
	}

	var isPbkdf2Used = false
	var pbkdf2IterationCount int
	var pbkdf2kdfHashFunction func() hash.Hash

	switch resolveKdfType(keyDerivationFunction) {
	case "pbkdf2":
		isPbkdf2Used = true
		pbkdf2IterationCount, pbkdf2kdfHashFunction = parsePbkdf2(keyDerivationFunction)
		break
	default:
		errorLog.Fatalln(fmt.Sprintf("The specified KDF \"%s\" is not supported", keyDerivationFunction))
	}

	var derivativeKey []byte

	if isPbkdf2Used {
		derivativeKey = pbkdf2.Key(encryptionKeyBytes, encryptionKeySaltBytes, pbkdf2IterationCount, keyLengthForAlgo(encryptionAlgorithm), pbkdf2kdfHashFunction)
	}

	if verbose {
		outLog.Println(fmt.Sprintf("key=%s", hex.EncodeToString(encryptionKeyBytes)))
		outLog.Println(fmt.Sprintf("salt=%s", hex.EncodeToString(encryptionKeySaltBytes)))
		outLog.Println(fmt.Sprintf("dk=%s", hex.EncodeToString(derivativeKey)))
	}

	var isAes256CbcUsed = false

	switch resolveAlgoType(encryptionAlgorithm) {
	case "aes256cbc":
		isAes256CbcUsed = true
		break
	}

	var plainBytes []byte
	var plainText string

	if isAes256CbcUsed {
		block, cipherError := aes.NewCipher(derivativeKey)
		if cipherError != nil {
			panic(cipherError)
		}

		plainBytes = make([]byte, len(cipherBytes))

		mode := cipher.NewCBCDecrypter(block, initializationVector)
		mode.CryptBlocks(plainBytes, cipherBytes[0:len(cipherBytes)])

		plainBytes, _ = pkcs7.Pkcs7strip(plainBytes, aes.BlockSize)

		plainText = string(plainBytes)
	}

	fmt.Printf("%s\n", plainText)
}

func keyLengthForAlgo(algo string) int {
	switch algo {
	case "aes256cbc":
		return 32
		break
	}

	return 32
}

func isSupportedAlgo(algo string) bool {
	switch algo {
	case "aes256cbc":
		return true
		break
	}

	return false
}

func resolveAlgoType(algo string) string {
	switch algo {
	case "aes256cbc":
		return "aes256cbc"
		break
	}

	return ""
}

func resolveKdfType(kdf string) string {
	if strings.HasPrefix(kdf, "pbkdf2") {
		return "pbkdf2"
	}

	return ""
}

func parsePbkdf2(kdf string) (int, func() hash.Hash) {
	var params = strings.Split(kdf, ":")

	if len(params) != 3 {
		panic("The amount of arguments for PBKDF2 are invalid")
	}

	var iterCount, _ = strconv.Atoi(params[1])

	var hashFunction func() hash.Hash

	switch params[2] {
	case "sha1":
		hashFunction = sha1.New
	case "sha256":
		hashFunction = sha256.New
	case "sha256/224":
		hashFunction = sha256.New224
	case "sha512/384":
		hashFunction = sha512.New384
	case "sha512/224":
		hashFunction = sha512.New512_224
	case "sha512/256":
		hashFunction = sha512.New512_256
	case "sha512":
		hashFunction = sha512.New
	default:
		panic(fmt.Sprintf("The invalid hash algorithm %s for PBKDF2", kdf))
	}

	return iterCount, hashFunction
}


enka_encryption/encryption.go
//
// Copyright (c) 2023 Joel Strasser <joelstrasser1@gmail.com>
//
// Licensed under the EUPL-1.2 license.
//
// For the full license text consult the 'LICENSE' file from the repository.
//

package enka_encryption

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/base64"
	"encoding/hex"
	"enka/pkcs7"
	"flag"
	"fmt"
	"golang.org/x/crypto/pbkdf2"
	"hash"
	"io"
	"log"
	"strconv"
	"strings"
)

var encryptionAlgorithm string
var keyDerivationFunction string
var encryptionKeySalt string
var encryptionKey string
var plainText string
var verbose bool

func Encrypt(args []string, outLog *log.Logger, errorLog *log.Logger) {
	fs := flag.NewFlagSet("encrypt", flag.ExitOnError)
	fs.StringVar(&encryptionAlgorithm, "algo", "aes256cbc", "The encryption algorithm to use")
	fs.StringVar(&keyDerivationFunction, "kdf", "pbkdf2:650000:sha256", "The KDF to use")
	fs.StringVar(&encryptionKey, "key", "", "The encryption key")
	fs.StringVar(&encryptionKeySalt, "salt", "", "A salt for the KDF")
	fs.StringVar(&plainText, "text", "", "The string to encrypt")
	fs.BoolVar(&verbose, "verbose", false, "Set verbosity on")
	err := fs.Parse(args)
	if err != nil {
		return
	}

	if !isSupportedAlgo(encryptionAlgorithm) {
		errorLog.Fatalln(fmt.Sprintf("The specified algorithm \"%s\" is not supported", encryptionAlgorithm))
	}

	var isPbkdf2Used = false
	var pbkdf2IterationCount int
	var pbkdf2kdfHashFunction func() hash.Hash

	switch resolveKdfType(keyDerivationFunction) {
	case "pbkdf2":
		isPbkdf2Used = true
		pbkdf2IterationCount, pbkdf2kdfHashFunction = parsePbkdf2(keyDerivationFunction)
		break
	default:
		errorLog.Fatalln(fmt.Sprintf("The specified KDF \"%s\" is not supported", keyDerivationFunction))
	}

	var encryptionKeyBytes = []byte(encryptionKey)
	var encryptionKeySaltBytes []byte

	if encryptionKeySalt == "" {
		encryptionKeySaltBytes = make([]byte, 8)
		_, saltError := rand.Read(encryptionKeySaltBytes)
		if saltError != nil {
			panic(saltError)
		}
	} else {
		encryptionKeySaltBytes = []byte(encryptionKeySalt)
	}

	var derivativeKey []byte

	if isPbkdf2Used {
		derivativeKey = pbkdf2.Key(encryptionKeyBytes, encryptionKeySaltBytes, pbkdf2IterationCount, keylengthForAlgo(encryptionAlgorithm), pbkdf2kdfHashFunction)
	}

	if verbose {
		outLog.Println(fmt.Sprintf("key=%s", hex.EncodeToString(encryptionKeyBytes)))
		outLog.Println(fmt.Sprintf("salt=%s", hex.EncodeToString(encryptionKeySaltBytes)))
		outLog.Println(fmt.Sprintf("dk=%s", hex.EncodeToString(derivativeKey)))
	}

	var isAes256CbcUsed = false

	switch resolveAlgoType(encryptionAlgorithm) {
	case "aes256cbc":
		isAes256CbcUsed = true
		break
	}

	var plainBytes []byte
	var cipherBytes []byte
	var initializationVector []byte

	if isAes256CbcUsed {
		block, cipherError := aes.NewCipher(derivativeKey)
		if cipherError != nil {
			panic(cipherError)
		}

		plainBytes = []byte(plainText)
		plainBytes, _ = pkcs7.Pkcs7pad(plainBytes, aes.BlockSize)

		cipherBytes = make([]byte, len(plainBytes))

		initializationVector = make([]byte, aes.BlockSize)
		if _, ivError := io.ReadFull(rand.Reader, initializationVector); ivError != nil {
			panic(ivError)
		}

		if verbose {
			outLog.Println(fmt.Sprintf("iv=%s", hex.EncodeToString(initializationVector)))
		}

		mode := cipher.NewCBCEncrypter(block, initializationVector)
		mode.CryptBlocks(cipherBytes, plainBytes)
	}

	fmt.Printf("%%enka%%v1%%%s%%%s%%%s%%%s%%%s\n", encryptionAlgorithm, keyDerivationFunction, base64.StdEncoding.EncodeToString(encryptionKeySaltBytes), base64.StdEncoding.EncodeToString(initializationVector), base64.StdEncoding.EncodeToString(cipherBytes))
}

func keylengthForAlgo(algo string) int {
	switch algo {
	case "aes256cbc":
		return 32
		break
	}

	return 32
}

func isSupportedAlgo(algo string) bool {
	switch algo {
	case "aes256cbc":
		return true
		break
	}

	return false
}

func resolveAlgoType(algo string) string {
	switch algo {
	case "aes256cbc":
		return "aes256cbc"
		break
	}

	return ""
}

func resolveKdfType(kdf string) string {
	if strings.HasPrefix(kdf, "pbkdf2") {
		return "pbkdf2"
	}

	return ""
}

func parsePbkdf2(kdf string) (int, func() hash.Hash) {
	var params = strings.Split(kdf, ":")

	if len(params) != 3 {
		panic("The amount of arguments for PBKDF2 are invalid")
	}

	var iterCount, _ = strconv.Atoi(params[1])

	var hashFunction func() hash.Hash

	switch params[2] {
	case "sha1":
		hashFunction = sha1.New
	case "sha256":
		hashFunction = sha256.New
	case "sha256/224":
		hashFunction = sha256.New224
	case "sha512/384":
		hashFunction = sha512.New384
	case "sha512/224":
		hashFunction = sha512.New512_224
	case "sha512/256":
		hashFunction = sha512.New512_256
	case "sha512":
		hashFunction = sha512.New
	default:
		panic(fmt.Sprintf("The invalid hash algorithm %s for PBKDF2", kdf))
	}

	return iterCount, hashFunction
}
